src/librustc/hir/lowering.rs

```diff
@@ -1149,13 +1149,69 @@ impl<'a> LoweringContext<'a> {
         }
     }
 
-    fn lower_generic_arg(&mut self,
-                         arg: &ast::GenericArg,
-                         itctx: ImplTraitContext<'_>)
-                         -> hir::GenericArg {
+    fn lower_generic_arg(
+        &mut self,
+        arg: &ast::GenericArg,
+        itctx: ImplTraitContext<'_>
+    ) -> hir::GenericArg {
         match arg {
             ast::GenericArg::Lifetime(lt) => GenericArg::Lifetime(self.lower_lifetime(&lt)),
-            ast::GenericArg::Type(ty) => GenericArg::Type(self.lower_ty_direct(&ty, itctx)),
+            ast::GenericArg::Type(ty) => {
+                // A path type argument could be an incorrectly parsed generic const argument. If
+                // we resolved the path in the value namespace, we transform it into a generic
+                // const argument.
+                if let TyKind::Path(ref qself, ref path) = ty.kind {
+                    if path.segments.len() == 1 {
```

Comment:
A comment here explaining this check would be helpful (as consts _can_ have multiple segments, we just can't allow them here).

Reply:
What exactly is the reasoning behind only allowing single segment consts here? I'm not sure I've ever seen the reason spelled out.

Reply:
Lowering doesn't care about this condition and can act on the resolution results only (multi-segment paths will never result in values right now, and if they will in the future, they'll need to be treated as consts too), so this line can be removed.